README.rst

```diff
@@ -1,103 +1,262 @@
-call ``createParser(list, size: uint16)`` would create a parser
-``proc (stream: Stream, size: uint16): <return type>``). To call a parser
-use the ``*`` type as described above and give it the name of the parser and
-any optional arguments. The stream object will get added automatically as the
-first parameter.
-
-When creating a parser you get a tuple with two members, ``get`` and ``put``
-which is stored by a let as the identifier given when calling createParser.
-These are both procedures, the first only takes a stream (and any optional
-arguments as described above) and returns a tuple containing all the fields.
-The second takes a stream and a tuple containing all the fields, this is the
-same tuple returned by the ``get`` procedure and writes the format to the
-stream.
+This module implements a DSL for creating binary parsers/encoders.
+It exports the macro ``createParser`` which generates a ``tuple[get: proc, put: proc]``.
+``get`` returns a tuple with each parsed field and ``put`` writes a compatible tuple to the stream.
+
+The macro accepts 3 kind of things:
+
+-  Parser options
+-  Parser parameters
+-  Block with DSL statements
+
+Parser options
+--------------
+
+Each specified option must be in the form ``option = value``:
+
+- ``endian``: sets the default endianness for the whole parser
+   - *default*: big endian
+   - ``b``: **big** endian
+   - ``l``: **little** endian
+- ``bitEndian``: sets the default bit-endianness for the whole parser
+   - *default*: left -> right
+   - ``n``: left -> right (**normal**)
+   - ``r``: left <- right (**reverse**)
+
+Parser parameters
+-----------------
+
+Each parameter must be in the form ``symbol: type``. The generated
+``get``/``put`` procs will then have this additional parameter appended.
+
+DSL
+----
+
+Each statement corresponds to 1 field. The general syntax is:
+
+.. code:: nim
+
+    Type {Operations}: Value
+
+where ``{Operations}`` is optional and refers to a plugin system (see
+below).
+
+Type
+~~~~
+
+The **kind**, **endianness** and **size** are encoded in a identifier
+made up of:
+
+- 1 optional letter specifying the kind:
+   - *default*: signed integer
+   - ``u``: **unsigned** integer
+   - ``f``: **float**
+   - ``s``:**string**
+   - ``*``: complex (see below)
+- 1 optional letter specifying endianness:
+   - *default*: big endian
+   - ``b``: **big** endian
+   - ``l``: **little** endian
+- 1 optional letter specifying big-endianness for unaligned reads:
+   - *default*: left -> right
+   - ``n``: left -> right (**normal**)
+   - ``r``: left <- right (**reverse**)
+- 1 number specifying size in **bits**:
+   - for strings only byte multiples are allows (``8``, ``16``, ``24``, ...)
+   - for an integer the allowed values are ``1 .. 64``
+   - for a float the allowed values are ``32`` and ``64``
+   - for a custom it can't be used (use the secondary ``size`` operation)
+
+| Strings may not have a size, in which case they are null-terminated.
```

Comment:
Maybe mention that fixed length strings are typically done by `s: myString[10]` so people don't do `s80: myString`

Reply:
I don't get this. In new implementation `s: myString[10]` means "10 null-terminated strings" while `s80: myString` means "1 80-bit string".

Reply:
Fixed length strings are done with substreams with the following syntax `s: myString(10)`.